### DISCUSSION TITLE
Page properties extraction

### DISCUSSION THREAD

**Discussion by meghanaviyyapu** [pymupdf/PyMuPDF#1871] [Looking for help] [answered]
Can you let me know how page margin values (left,top,right,bottom) can be extracted from PDF?

**Comment by JorjMcKie** [comment, marked as answer]
There is no such thing.
You can however determine the union of all text bboxes easily by computing them e.g. from `get_text("blocks")`.
Or build other types of sub areas in similar ways (drawings, images, ...).

**Reply by meghanaviyyapu** [reply]
If the y1 Rect value of the first character in a PDF is taken and subtracted from page height, does it help in getting page top margin value?

**Reply by JorjMcKie** [reply]
No.
First you need to confirm that the extracted text is sorted in reading order.
I suggest to do this instead (which is extremely fast!):
```python
margins=fitz.EMPTY_RECT()
for b in page.get_text("blocks", flags=fitz.TEXTFLAGS_TEXT):
    margins |= b[:4]
# now x0, y0, x1, y1 of rectangle margins are the values you want.
```

**Reply by meghanaviyyapu** [reply]
Thank you for explaining.
I have taken an input PDF with (left 36pt, right 72pt, top 108pt, bottom 180pt) margin values and extracted the values with the above code. It gave me different values. Can you let me know if any transformation has to be applied?

[margin1.pdf](https://github.com/pymupdf/PyMuPDF/files/9314395/margin1.pdf)
![Capture](https://user-images.githubusercontent.com/36301306/184303300-5ac8f2f9-8ee1-45da-894a-f8a8f51b7a3d.JPG)

**Reply by JorjMcKie** [reply]
1. Margins, as you understand them, **_do not exist in a PDF_**. There is no place where they are stored. PDF only knows its different rectangles like MediaBox, CropBox, etc. None of them is anything close to what you want.
2. The only thing coming close to your understanding are the bboxes of objects on a page (like text).
3. What you computed here, is the joined area of text objects. This bbox lies inside of the rectangle that you defined by your margin values. This because you inserted the text like that.

**Reply by meghanaviyyapu** [reply]
I have tried extracting different rectangles but those values are not closer to what I am expecting. Can you let me know if there is any way to get page margin values?

**Reply by JorjMcKie** [reply]
> Can you let me know if there is any way to get page margin values?

No - because **_there is no such thing like a margin_**.

**Reply by meghanaviyyapu** [reply]
Ok thank you. Is it possible to calculate them instead of getting them from the internal structure?

**Reply by JorjMcKie** [reply]
Again, no.
**_Margins do not exist,_** nowhere.
Except in your script, maybe.

**Reply by meghanaviyyapu** [reply]
Ok thank you

**Reply by meghanaviyyapu** [reply]
![image](https://user-images.githubusercontent.com/36301306/188084238-e89596e3-221e-4c38-9ea0-d68fce27e041.png)

This code is extracting accurate left margin values but not other margin values. Can you let me know the reason for this?

**Reply by JorjMcKie** [reply]
Well, this is one piece of info of the middle of presumably many other pieces. So it is hard to see what the problem may be - if there is any.

Then I don't understand, why you use `page.mediabox` instead of `page.rect`.

**Reply by meghanaviyyapu** [reply]
page.mediabox and page.rect gave the same values

**Reply by JorjMcKie** [reply]
Then, according to your logic the bottom "margin" is margins.y1 (or rather the difference page.rect.height margins.y1). And top is margins.y0.

I cannot image why you would want this info at all. But given you need it, you should probably build the union of text rectangles for every page and take the maximum rect (i.e. the union) of all those across all pages.

**Reply by meghanaviyyapu** [reply]
I have modified the logic for top and bottom margins. The actual margin values of the input PDF are  (left 36pt, right 72pt, top 108pt, bottom 180pt) but only the left margin value seems to be accurate using the below logic.
![image](https://user-images.githubusercontent.com/36301306/188316065-14013d9f-cb9b-46be-b40b-dbce4eda8d5c.png)
[margin1.pdf](https://github.com/pymupdf/PyMuPDF/files/9484791/margin1.pdf)


Can a PDF be created from an existing PDF without specifying margin values?

**Reply by JorjMcKie** [reply]
> The actual margin values of the input PDF are (left 36pt, right 72pt, top 108pt, bottom 180pt)

Who told you that? In PDF, there exists no such concept like a "margin".

> Can a PDF be created from an existing PDF without specifying margin values?

Of course, because a PDF page has no such thing like a "margin". Such values are not stored anywhere in a PDF document. So you cannot find them. The best approximation you can ever get, is what I explained already.
If you believe you have better insight, please tell me where to find such values.

> only the left margin value seems to be accurate

No, **_all_** values are correct!

**Reply by meghanaviyyapu** [reply]
I found a code to create a PDF with margins using a library called OpenPDF.
        Document document = new Document(PageSize.A5, 36, 72, 108, 180);
Here 36, 72, 108, 180 are margin values. If a PDF is created without specifying these values, will there be any effect on alignment of text?

**Reply by JorjMcKie** [reply]
> without specifying these values, will there be any effect on alignment of text?

Presumably yes.
But this does not mean that you can extract those values from the PDF again once it has been created!
